src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,593 @@
+            }
+            set
+            {
+                this.securityAuthChannelDispatcher = value;
+            }
+        }
+
+        // public EndpointDispatcherTable Endpoints => ChannelDispatcher.EndpointDispatcherTable;
+
+        public ICollection<Type> SupportedChannelTypes => InnerServiceDispatcher.SupportedChannelTypes;
+
+        public object ThisLock { get; } = new object();
+
+        public SecurityProtocolFactory SecurityProtocolFactory
+        {
+            get
+            {
+
+                //  ThrowIfDisposed();
+                return this.securityProtocolFactory;
+            }
+            set
+            {
+                if (value == null)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("value");
+                }
+                //   ThrowIfDisposedOrImmutable();
+                this.securityProtocolFactory = value;
+            }
+        }
+
+        public bool SessionMode
+        {
+            get
+            {
+                return this.sessionMode;
+            }
+            set
+            {
+                this.sessionMode = value;
+            }
+        }
+
+        public SecuritySessionServerSettings SessionServerSettings
+        {
+            get
+            {
+                if (this.sessionServerSettings == null)
+                {
+                    lock (ThisLock)
+                    {
+                        if (this.sessionServerSettings == null)
+                        {
+                            SecuritySessionServerSettings tmp = new SecuritySessionServerSettings();
+                            System.Threading.Thread.MemoryBarrier();
+                            tmp.SecurityServiceDispatcher = this;
+                            this.sessionServerSettings = tmp;
+                        }
+                    }
+                }
+                return this.sessionServerSettings;
+            }
+        }
+
+        bool SupportsDuplex
+        {
+            get
+            {
+                ThrowIfProtocolFactoryNotSet();
+                return this.securityProtocolFactory.SupportsDuplex;
+            }
+        }
+
+        bool SupportsRequestReply
+        {
+            get
+            {
+                ThrowIfProtocolFactoryNotSet();
+                return this.securityProtocolFactory.SupportsRequestReply;
+            }
+        }
+
+        public bool SendUnsecuredFaults
+        {
+            get
+            {
+                return this.sendUnsecuredFaults;
+            }
+            set
+            {
+                this.sendUnsecuredFaults = value;
+            }
+        }
+
+        public IChannel OuterChannel
+        {
+            get { return this.outerChannel; }
+            set { this.outerChannel = value; }
+        }
+
+        IList<Type> IServiceDispatcher.SupportedChannelTypes => throw new NotImplementedException();
+
```

Comment:
You can implement this. Currently securityProtocolFactory has two methods, SupportsRequestReply and SupportsDuplex. It has an implementation internally to calculate those things. You could add a SupportedChannelTypes property to SecurityProtocolFactory which expresses this information in the form of channel shapes (e.g. IReplyChannel, IDuplexSessionChannel etc).